core/src/main/java/io/grpc/internal/TransportSet.java

```diff
@@ -106,10 +110,11 @@
   private boolean shutdown;
 
   /**
-   * The future for the transport for new outgoing requests. 'lock' lock must be held when assigning
-   * to activeTransportFuture.
+   * The future for the transport for new outgoing requests.
    */
-  private volatile SettableFuture<ClientTransport> activeTransportFuture;
+  @Nullable
+  @GuardedBy("lock")
+  private SettableFuture<ClientTransport> activeTransportFuture;
```

Comment:
Could this go back to being `volatile` and only acquire the lock when needing to update the field? That is pretty similar (but should be easier) to what was done for `ChannelImpl` in 9a2db9d. That keeps the fast path lock-free.

```java
SettableFuture<ClientTransport> savedActiveTransportFuture = activeTransportFuture;
if (savedActiveTransportFuture != null) {
  return savedActiveTransportFuture;
}
synchronized (lock) {
  // Check again, since it could have changed before acquiring the lock
  savedActiveTransportFuture = activeTransportFuture;
  if (savedActiveTransportFuture != null) {
    return savedActiveTransportFuture;
  }
  createActiveTransportFuture();
  return Preconditions.checkNotNull(activeTransportFuture, "activeTransportFuture");
}